File: src/CountingBits.java
public class CountingBits {
    public static int[] countBits(int n) {
        int[] result = new int[n + 1];
        if (result.length > 1) result[1] = 1;
        for (int k = 2 ; k < result.length ; k *= 2) {
            for (int i = 0 ; i < k && i + k < result.length; i++) {
                result[k + i] = 1 + result[i];
            }
        }
        return result;
    }
}

File: python/counting_bits.py
from typing import List


class Solution:
    def countBits(self, n: int) -> List[int]:
        result = [0] * (n + 1)
        if len(result) > 1: result[1] = 1
        i, k = 0, 2
        while k < len(result):
            i = 0
            while i < k and i + k < len(result):
                result[i + k] = 1 + result[i]
                i += 1
            k *= 2
        return result
